src/lib.rs

```diff
@@ -71,15 +71,78 @@ impl WaiterChannel {
     }
 }
 
-struct Scheduler {
-    waiter: Arc<WaiterChannel>,
-    heap: BinaryHeap<Schedule>,
+/// A trait that allows configurable execution of scheduled item
+/// on the scheduler thread.
+trait Executor<T> {
+    // Due to difference in use between Box<FnMut()> and most other data
+    // types, this trait requires implementors to provide two implementations
+    // of execute. While both of these functions execute the data item
+    // they differ on whether they make an equivalent data item available
+    // to the Scheduler to store in recurring schedules.
+    //
+    // execute() is called whenever a non-recurring data item needs
+    // to be executed, and consumes the data item in the process.
+    //
+    // execute_clone() is called whenever a recurring data item needs
+    // to be executed, and produces a new equivalent data item. This
+    // function should be more or less equivalent to:
+    //
+    // fn execute_clone(&mut self, data : T) -> T {
+    //   self.execute(data.clone());
+    //   data
+    // }
+
+    fn execute(&mut self, data : T);
+
+    fn execute_clone(&mut self, data : T) -> T;
```

Comment:
Wouldn't a `fn execute_ref(&mut self, data: &T)` have served the same purpose with a slightly simpler API? I'm not asking you to change this, btw.